File: apps/web/src/lib/capture-quality.test.ts
import { describe, expect, it } from 'vitest';

import {
	assessCapture,
	laplacianVariance,
	SMALL_TEXT_HEIGHT,
	SOFT_IMAGE_LAPLACIAN_VARIANCE
} from './capture-quality';
import { MIN_LINE_HEIGHT } from './segmentation';

/** White page with full-height dark bars in the given row bands. */
function page(
	width: number,
	height: number,
	bands: [number, number][],
	pitch = 8,
	ink = 0
): ImageData {
	const data = new Uint8ClampedArray(width * height * 4);
	for (let i = 0; i < width * height; i++) {
		const o = i * 4;
		data[o] = data[o + 1] = data[o + 2] = 255;
		data[o + 3] = 255;
	}
	for (const [y0, y1] of bands) {
		for (let y = y0; y < y1; y++) {
			for (let x = 20; x < width - 20; x += pitch) {
				const o = (y * width + x) * 4;
				data[o] = data[o + 1] = data[o + 2] = ink;
			}
		}
	}
	return { width, height, data } as ImageData;
}

describe('laplacianVariance', () => {
	it('a flat field has no edges', () => {
		expect(laplacianVariance(page(60, 60, []))).toBe(0);
	});

	it('a hard-edged page scores far above a flat one', () => {
		const sharp = laplacianVariance(page(200, 200, [[80, 120]]));
		expect(sharp).toBeGreaterThan(SOFT_IMAGE_LAPLACIAN_VARIANCE);
	});

	it('softening the same content lowers the score', () => {
		const hard = page(200, 200, [[80, 120]]);
		// Blur by hand: replace hard black with mid grey, which halves the step at
		// every edge without moving any edge.
		const soft = page(200, 200, [[80, 120]], 8, 128);
		expect(laplacianVariance(soft)).toBeLessThan(laplacianVariance(hard));
	});

	/**
	 * Pins the kernel coefficient, not just the ordering.
	 *
	 * A single grey pixel on white has an analytically known response, and a grey
	 * centre is required: a black centre gives the same answer under -4 and -3
	 * because it is multiplied by zero either way.
	 *
	 * 5x5, centre 100 on 255. Interior is 3x3 = 9 samples. The centre responds
	 * 4*255 - 4*100 = 620; its four edge-neighbours each 3*255 + 100 - 4*255 = -155;
	 * the four corners 0. Mean is (620 - 620)/9 = 0, so the variance is the mean
	 * square: (620^2 + 4*155^2)/9 = 480500/9.
	 */
	it('matches the analytic response of a known kernel', () => {
		const w = 5;
		const h = 5;
		const data = new Uint8ClampedArray(w * h * 4);
		for (let i = 0; i < w * h; i++) {
			const o = i * 4;
			data[o] = data[o + 1] = data[o + 2] = 255;
			data[o + 3] = 255;
		}
		const centre = (2 * w + 2) * 4;
		data[centre] = data[centre + 1] = data[centre + 2] = 100;

		expect(laplacianVariance({ width: w, height: h, data } as ImageData)).toBeCloseTo(
			480500 / 9,
			2
		);
	});

	it('an image too small to hold the kernel scores zero rather than throwing', () => {
		expect(laplacianVariance(page(2, 2, []))).toBe(0);
		expect(laplacianVariance(page(1, 40, []))).toBe(0);
	});
});

describe('assessCapture', () => {
	it('a clean page of ordinary text has nothing to warn about', () => {
		const clean = page(600, 400, [
			[60, 110],
			[160, 210],
			[260, 310]
		]);
		const result = assessCapture(clean);
		expect(result.lineCount).toBeGreaterThan(0);
		expect(result.medianLineHeight).toBeGreaterThanOrEqual(SMALL_TEXT_HEIGHT);
		expect(result.warnings).toEqual([]);
	});

	it('warns when no lines were found at all', () => {
		const blank = page(200, 200, []);
		const result = assessCapture(blank);
		expect(result.lineCount).toBe(0);
		expect(result.warnings.join(' ')).toContain('No text lines were found');
	});

	/**
	 * Pins the core-to-reported factor that `SMALL_TEXT_HEIGHT` is derived from. A
	 * band one pixel above the drop threshold reports far taller than it is drawn,
	 * because vertical smearing widens the run and padding is added either side. If
	 * this number moves, `SMALL_TEXT_HEIGHT` no longer means what its comment says.
	 */
	it('reportedHeightOfABarelyKeptBand', () => {
		const core = MIN_LINE_HEIGHT + 1;
		const p = page(600, 300, [[40, 40 + core]]);
		const result = assessCapture(p);
		expect(result.lineCount).toBe(1);
		expect(result.medianLineHeight).toBe(27);
		// Which is what makes 3x the drop threshold the right place for the warning.
		expect(result.medianLineHeight).toBeLessThanOrEqual(SMALL_TEXT_HEIGHT);
	});

	it('warns about text close to the silent-drop threshold', () => {
		// Bands of MIN_LINE_HEIGHT + 1: kept, but only just.
		const h = MIN_LINE_HEIGHT + 1;
		const tiny = page(600, 300, [
			[40, 40 + h],
			[120, 120 + h],
			[200, 200 + h]
		]);
		const result = assessCapture(tiny);
		const text = result.warnings.join(' ');
		expect(result.lineCount).toBeGreaterThan(0);
		expect(text).toContain('discarded without notice');
		// The warning must name the real threshold, not a copy of it.
		expect(text).toContain(String(MIN_LINE_HEIGHT));
	});

	it('warns that a soft image will lose diacritics', () => {
		// A flat field has no edges at all, so it is the extreme of soft. It also has
		// no lines, so both warnings are expected — this asserts the soft one is
		// present, which is what the threshold controls.
		const result = assessCapture(page(120, 120, []));
		expect(result.sharpness).toBeLessThan(SOFT_IMAGE_LAPLACIAN_VARIANCE);
		expect(result.warnings.join(' ')).toContain('soft');
	});

	it('reports the sharpness it measured', () => {
		const clean = page(400, 300, [[100, 160]]);
		expect(assessCapture(clean).sharpness).toBeCloseTo(laplacianVariance(clean), 6);
	});

	it('warns about bands that are not line-shaped', () => {
		// One squarish block filling most of a small page: fails the aspect test and
		// exceeds the page fraction, so `looksLikeALine` rejects it.
		const w = 300;
		const h = 300;
		const data = new Uint8ClampedArray(w * h * 4);
		for (let i = 0; i < w * h; i++) {
			const o = i * 4;
			data[o] = data[o + 1] = data[o + 2] = 255;
			data[o + 3] = 255;
		}
		for (let y = 30; y < 270; y++) {
			for (let x = 30; x < 270; x += 6) {
				const o = (y * w + x) * 4;
				data[o] = data[o + 1] = data[o + 2] = 0;
			}
		}
		const result = assessCapture({ width: w, height: h, data } as ImageData);
		if (result.lineCount > 0) {
			expect(result.warnings.join(' ')).toContain('not line-shaped');
		}
	});
});

File: apps/android/app/src/test/java/dev/janakhpon/monocr/engine/CaptureQualityTest.kt
package dev.janakhpon.monocr.engine

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import kotlin.math.abs

/**
 * Sharpness measurement.
 *
 * The analytic case is the load-bearing one: it pins the kernel COEFFICIENT, not just
 * the ordering of two images. A test that only asserts "sharp scores higher than soft"
 * passes with the wrong coefficient, which is how the same test on the web port
 * survived a mutation from -4 to -3 until it was tightened.
 *
 * These mirror `apps/ios/MonOcrCore/Tests/MonOcrCoreTests/CaptureQualityTests.swift`
 * case for case and use the same derived values, so the three ports are pinned to one
 * derivation rather than three.
 *
 * Two mutations of [CaptureQuality] survive this suite and are not oversights:
 *
 *  - Relaxing the 3x3 minimum to 2x2 changes nothing observable. The loops already
 *    run `1 until size - 1`, so a 2-pixel axis yields no samples and the function
 *    returns 0 through the `n == 0` path instead of the guard. The guard states the
 *    filter's requirement rather than implementing it.
 *  - `isSoft`'s `<` against `<=` differs only where the variance is exactly the
 *    floor. `20 * d^2 / 9` is the variance of the 5x5 family used above and no
 *    integer `d` makes it 100, so there is no cheap image that lands on the
 *    boundary; the floor is a documented heuristic rather than a measured value, so
 *    contriving one would pin a number that is meant to move.
 */
class CaptureQualityTest {

    /** White page with dark bars in one row band, [pitch] apart. */
    private fun page(
        width: Int, height: Int, band: IntRange?, pitch: Int = 8, ink: Int = 0
    ): GreyImage {
        val px = IntArray(width * height) { 255 }
        if (band != null) {
            for (y in band) {
                var x = 20
                while (x < width - 20) {
                    px[y * width + x] = ink
                    x += pitch
                }
            }
        }
        return GreyImage(width, height, px)
    }

    /**
     * The analytic case, derived by hand and reproduced independently on two ports.
     *
     * 5x5, all 255, centre 100. A grey centre is required: a black centre gives the
     * same answer under -4 and -3 because it is multiplied by zero either way, so it
     * cannot pin the coefficient.
     *
     * Interior is 3x3 = 9 samples. The centre responds 4*255 - 4*100 = 620; its four
     * edge-neighbours each 3*255 + 100 - 4*255 = -155; the four interior corners 0.
     * The sum is 620 - 620 = 0 exactly, so the mean is 0 and the variance is the mean
     * square: (620^2 + 4*155^2)/9 = 480500/9.
     */
    @Test
    fun `matches the analytic response of a known kernel`() {
        val px = IntArray(25) { 255 }
        px[2 * 5 + 2] = 100
        val variance = CaptureQuality.laplacianVariance(GreyImage(5, 5, px))
        assertTrue(
            "expected 480500/9, got $variance",
            abs(variance - 480500.0 / 9.0) < 1e-9
        )
    }

    @Test
    fun `a flat field has no edges`() {
        assertEquals(0.0, CaptureQuality.laplacianVariance(page(60, 60, null)), 0.0)
    }

    @Test
    fun `a hard edged page scores above the soft floor`() {
        val sharp = CaptureQuality.laplacianVariance(page(200, 200, 80 until 120))
        assertTrue(
            "a page of hard bars should not read as soft, got $sharp",
            sharp > CaptureQuality.SOFT_IMAGE_LAPLACIAN_VARIANCE
        )
    }

    /** Same content, half the step at every edge: no edge moves, only its contrast. */
    @Test
    fun `softening the same content lowers the score`() {
        val hard = CaptureQuality.laplacianVariance(page(200, 200, 80 until 120))
        val soft = CaptureQuality.laplacianVariance(page(200, 200, 80 until 120, ink = 128))
        assertTrue("soft=$soft should be under hard=$hard", soft < hard)
    }

    @Test
    fun `an image too small for the kernel scores zero rather than throwing`() {
        assertEquals(0.0, CaptureQuality.laplacianVariance(page(2, 2, null)), 0.0)
        assertEquals(0.0, CaptureQuality.laplacianVariance(page(1, 40, null)), 0.0)
        assertEquals(0.0, CaptureQuality.laplacianVariance(GreyImage(0, 0, IntArray(0))), 0.0)
    }

    @Test
    fun `softness is reported against the floor`() {
        assertTrue(CaptureQuality.isSoft(page(120, 120, null)))
        assertFalse(CaptureQuality.isSoft(page(200, 200, 80 until 120)))
    }

    /**
     * A constant response has zero variance even though its mean is large.
     *
     * This is the case that separates the variance from the mean square, and the
     * only one that does. Every other image here has a near-zero Laplacian mean —
     * the filter telescopes to boundary terms over the interior, which is exactly
     * why the single-pass `E[x^2] - E[x]^2` form is safe in this function. The cost
     * of that is that dropping the `- mean * mean` term changes almost nothing, and
     * a mutation doing so survived the whole suite.
     *
     * A quadratic ramp is the exception. `p(x) = 2x^2` has an exact second
     * difference of 4, so every interior pixel responds 4: the mean is 4 and the
     * spread is nothing. The variance is 0 and the mean square is 16.
     */
    @Test
    fun `a constant response has zero variance and a non-zero mean`() {
        val w = 11
        val h = 5
        // 2x^2 over 0..10 peaks at 200, inside the 8-bit range, and every row is
        // identical so the vertical term contributes nothing.
        val px = IntArray(w * h) { 2 * (it % w) * (it % w) }
        val variance = CaptureQuality.laplacianVariance(GreyImage(w, h, px))
        assertTrue(
            "a constant response must have no spread; got $variance, and 16.0 would " +
                "mean the mean was not subtracted",
            abs(variance) < 1e-9
        )
    }

    /** One value for one question, across all three ports. */
    @Test
    fun `the floor matches the other ports`() {
        assertEquals(100.0, CaptureQuality.SOFT_IMAGE_LAPLACIAN_VARIANCE, 0.0)
    }
}
